### PR TITLE
coldata: change Batch interface to operate with uint64 length

### PR DESCRIPTION
**Pull request by yuzefovich**
This commit changes Batch interface to operate with uint64 length
because it unifies the way we store batches - now we will have a single
implementation MemBatch (with an exception for coldata.ZeroBatch). This
allows us to remove bufferedBatch.

Release note: None

### CODE REVIEW

**Comment by yuzefovich** [comment]
This is RFAL.

I did most of this work here while investigating #44140 (initially, I thought that that issue had a different root cause, so I started changing the `Batch` interface). Now I started wondering though whether this change is beneficial - this does unify the way we store "big" (i.e. buffered) and "normal" batches at the expense of minor (I hope) increase in memory footprint since now selection vectors operate on `uint64` values.

If we do decide that this should be merged, I believe we will need to do minor adjustments to "unwrap" big batches into `colvecs` (similar to how `cfetcher` does things). That could be done during stabilization period though.

File: pkg/col/coldata/batch.go [hunk 1]
```diff
@@ -156,17 +156,17 @@ func (b *zeroBatch) Reset([]coltypes.T, int) {
...
 	// length of batch or sel in tuples
-	n uint16
+	n uint64
 	// slice of columns in this batch.
 	b      []Vec
 	useSel bool
 	// if useSel is true, a selection vector from upstream. a selection vector is
 	// a list of selected column indexes in this memBatch's columns.
-	sel []uint16
+	sel []uint64
```
Comment (asubiotto) [inline pkg/col/coldata/batch.go]: Hmm, I remember that jordan ran some benchmarks with this change and performance decreased significantly

Comment (yuzefovich) [reply pkg/col/coldata/batch.go]: I guess we should wait for Jordan with this PR.

**Comment by yuzefovich** [comment]
I ran some benchmarks for this change (https://gist.github.com/yuzefovich/76300028e5ef2ca3608fc645767826b8).

They didn't show any significant degradation (apart from memory allocation in some cases, but that was expected). So I'm curious whether we should move forward with this. @jordanlewis @asubiotto

**Comment by asubiotto** [comment]
Interesting. Taking into account the benchmarks results and the unification of uint16-64, I personally would want to go forward with this change (what if we make it an `int`?) but I'd like to get @jordanlewis' opinion on this.

**Comment by yuzefovich** [comment]
I talked with Jordan, and he's on board with the change. We decided it'd be worth to rerun the selection benchmark again which I did on the gceworker:
```
name                                                   old time/op    new time/op    delta
SelLTInt64Int64ConstOp/useSel=true,hasNulls=true-24      6.88µs ± 1%    6.39µs ± 2%   -7.10%  (p=0.000 n=10+10)
SelLTInt64Int64ConstOp/useSel=true,hasNulls=false-24     2.46µs ± 1%    2.61µs ± 1%   +6.10%  (p=0.000 n=9+10)
SelLTInt64Int64ConstOp/useSel=false,hasNulls=true-24     5.92µs ± 2%    6.24µs ± 1%   +5.28%  (p=0.000 n=10+9)
SelLTInt64Int64ConstOp/useSel=false,hasNulls=false-24    2.23µs ± 1%    1.71µs ± 1%  -23.21%  (p=0.000 n=10+10)
SelLTInt64Int64Op/useSel=true,hasNulls=true-24           11.1µs ± 2%    12.2µs ± 3%   +9.83%  (p=0.000 n=10+10)
SelLTInt64Int64Op/useSel=true,hasNulls=false-24          3.30µs ± 0%    2.86µs ± 1%  -13.24%  (p=0.000 n=10+9)
SelLTInt64Int64Op/useSel=false,hasNulls=true-24          10.7µs ± 3%    10.7µs ± 2%     ~     (p=0.393 n=10+10)
SelLTInt64Int64Op/useSel=false,hasNulls=false-24         2.72µs ± 2%    2.16µs ± 1%  -20.49%  (p=0.000 n=10+9)

name                                                   old speed      new speed      delta
SelLTInt64Int64ConstOp/useSel=true,hasNulls=true-24    1.19GB/s ± 1%  1.28GB/s ± 2%   +7.66%  (p=0.000 n=10+10)
SelLTInt64Int64ConstOp/useSel=true,hasNulls=false-24   3.33GB/s ± 1%  3.14GB/s ± 1%   -5.75%  (p=0.000 n=9+10)
SelLTInt64Int64ConstOp/useSel=false,hasNulls=true-24   1.38GB/s ± 2%  1.31GB/s ± 1%   -5.01%  (p=0.000 n=10+9)
SelLTInt64Int64ConstOp/useSel=false,hasNulls=false-24  3.68GB/s ± 1%  4.79GB/s ± 1%  +30.21%  (p=0.000 n=10+10)
SelLTInt64Int64Op/useSel=true,hasNulls=true-24         1.48GB/s ± 2%  1.35GB/s ± 4%   -8.94%  (p=0.000 n=10+10)
SelLTInt64Int64Op/useSel=true,hasNulls=false-24        4.97GB/s ± 0%  5.73GB/s ± 1%  +15.26%  (p=0.000 n=10+9)
SelLTInt64Int64Op/useSel=false,hasNulls=true-24        1.53GB/s ± 3%  1.53GB/s ± 2%     ~     (p=0.393 n=10+10)
SelLTInt64Int64Op/useSel=false,hasNulls=false-24       6.02GB/s ± 2%  7.57GB/s ± 1%  +25.76%  (p=0.000 n=10+9)
```
and the results look roughly the same as on my laptop. For some reason, this change actually gives noticeable performance improvement in some cases.

I think we're all on board the we should proceed forward with this change, so PTAL.

**Review by yuzefovich — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44176)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto)

___
*[Makefile, line 1504 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44176#-M1BKvOm0BmG13T2ffux:-M1BKvOnCRmsgWeY3FZz:bgn3w2w) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c0ac04cbd3db0b0ac4b877c830980e0332240d2/Makefile#L1504)):*
> ```Makefile
> pkg/sql/colexec/count_agg.eg.go: pkg/sql/colexec/count_agg_tmpl.go
> pkg/sql/colexec/distinct.eg.go: pkg/sql/colexec/distinct_tmpl.go
> pkg/sql/colexec/hash_aggregator.eg.go: pkg/sql/colexec/hash_aggregator_tmpl.go
> ```

cc @azhng when adding a new template, it needs to be added in two places (from my understanding, one is responsible for telling `make` which template to use when generating code, and another is responsible for specifying the "target" which - if changed - should trigger regeneration of the code).

**Review by jordanlewis — COMMENTED** [review]
Okay, I'm on board.

By the way, do we account for the `sel` vector in memory accounting...?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44176#-:-M1ByUce4f1V2CxX9jJF:b-c3hrof)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto)

**Comment by yuzefovich** [comment]
We do (at least we try to account for `sel`).

I opened up #45534 to change `uint16` to `int` instead of `uint64`, I'll probably close this PR but will keep the branch, just in case.

**Comment by yuzefovich** [comment]
Closing this in favor of #45534.

**Pull request closed without merging**